core/lib/zksync_core/src/api_server/execution_sandbox/vm_metrics.rs

```diff
@@ -1,104 +1,189 @@
 use zksync_state::StorageViewMetrics;
 use zksync_types::event::{extract_long_l2_to_l1_messages, extract_published_bytecodes};
 use zksync_types::fee::TransactionExecutionMetrics;
 use zksync_types::storage_writes_deduplicator::StorageWritesDeduplicator;
 use zksync_utils::bytecode::bytecode_len_in_bytes;
 
-pub fn report_vm_memory_metrics(
+use crate::metrics::InteractionType;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "type", rename_all = "snake_case")]
+enum SizeType {
+    Inner,
+    History,
+}
+
+const MEMORY_SIZE_BUCKETS: Buckets = Buckets::values(&[
+    1_000.0,
+    10_000.0,
+    100_000.0,
+    500_000.0,
+    1_000_000.0,
+    5_000_000.0,
+    10_000_000.0,
+    50_000_000.0,
+    100_000_000.0,
+    500_000_000.0,
+    1_000_000_000.0,
+]);
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "runtime_context_memory")]
+struct RuntimeContextMemoryMetrics {
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    event_sink_size: Family<SizeType, Histogram<usize>>,
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    memory_size: Family<SizeType, Histogram<usize>>,
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    decommitter_size: Family<SizeType, Histogram<usize>>,
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    storage_size: Family<SizeType, Histogram<usize>>,
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    storage_view_cache_size: Histogram<usize>,
+    #[metrics(buckets = MEMORY_SIZE_BUCKETS)]
+    full: Histogram<usize>,
+}
+
+#[vise::register]
+static MEMORY_METRICS: vise::Global<RuntimeContextMemoryMetrics> = vise::Global::new();
+
+const INTERACTION_AMOUNT_BUCKETS: Buckets = Buckets::exponential(10.0..=10_000_000.0, 10.0);
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "runtime_context_storage_interaction")]
+struct RuntimeContextStorageMetrics {
+    #[metrics(buckets = INTERACTION_AMOUNT_BUCKETS)]
+    amount: Family<InteractionType, Histogram<usize>>,
+    #[metrics(buckets = Buckets::LATENCIES)]
+    duration: Family<InteractionType, Histogram<Duration>>,
+    #[metrics(buckets = Buckets::LATENCIES)] // FIXME: probably should use smaller buckets
+    duration_per_unit: Family<InteractionType, Histogram<Duration>>,
+    #[metrics(buckets = Buckets::ZERO_TO_ONE)]
+    ratio: Histogram<f64>,
+}
+
+#[vise::register]
+static STORAGE_METRICS: vise::Global<RuntimeContextStorageMetrics> = vise::Global::new();
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "stage", rename_all = "snake_case")]
+pub(super) enum SandboxStage {
+    VmConcurrencyLimiterAcquire,
+    Initialization,
+    ValidateInSandbox,
+    Validation,
+    Execution,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "stage", rename_all = "snake_case")]
+pub(in crate::api_server) enum SubmitTxStage {
+    #[metrics(name = "1_validate")]
```

Comment:
Manual specs could be avoided if we don't need numeric prefixes.

Reply:
I don't think it's that bad; can be migrated on per-case basis later.